packages/core/src/node/process/process.ts

```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright (C) 2017 Ericsson and others.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+import { injectable } from "inversify";
+import { ILogger } from '../../common/logger';
+import * as child from 'child_process';
+import * as stream from 'stream';
+import { Emitter, Event } from '../../common/event';
+import { Disposable } from '../../common/disposable';
+
+export interface IProcessExitEvent {
+    code: number,
+    signal?: string
+}
+
+@injectable()
+export abstract class Process implements Disposable {
+
+    abstract readonly type: 'Raw' | 'Terminal';
+    abstract pid: number;
+    killed = false;
+    abstract output: stream.Readable;
+    protected abstract process: child.ChildProcess | undefined;
+    protected abstract terminal: any;
+    protected readonly exitEmitter = new Emitter<IProcessExitEvent>();
+    protected readonly errorEmitter = new Emitter<Error>();
+
+    constructor(protected readonly logger: ILogger) {
+        this.exitEmitter.event(this.handleOnExit.bind(this));
+        this.errorEmitter.event(this.handleOnError.bind(this));
+    }
+
+    abstract kill(signal?: string): void;
+
+    get onExit(): Event<IProcessExitEvent> {
+        return this.exitEmitter.event;
+
+    }
+
+    get onError(): Event<Error> {
+        return this.errorEmitter.event;
+
+    }
+
+    dispose() {
+        if (this.killed === false) {
+            const p = new Promise<void>(resolve => {
+                this.kill();
+                this.exitEmitter.event(event => resolve());
+            });
+            p.then(() => {
+                this.exitEmitter.dispose();
+                this.errorEmitter.dispose();
+            });
+        } else {
+            this.exitEmitter.dispose();
+            this.errorEmitter.dispose();
+        }
+    }
+
+    protected emitOnExit(code: number, signal?: string) {
+        this.exitEmitter.fire({ 'code': code, 'signal': signal });
```

Comment:
What if `handleOn*` methods are called directly from `emitOn*`? then there should not be a cycle between emitters and the process and `dispose` method is not necessary?
Disposing of emitters also could be scheduled in `emitOn*` via `setTimeout` that existing clients could receive events but don't need to call `dispose` explicitly.

Reply:
Dispose would still be necessary to cleanup other external event emitters.
I'm not sure if we should always leave it to event listeners to clean up properly. I tried that in the test before and it leads to code like:

```
const sub = term.onExit(() => {
// something
  sub.dispose();
}
```
Each onSomthing needs to be handled.... rather then only the term object. I think it's quite error prone.

I for disposing with setTimeout (and I guess it should be process.nextTick )... it would be ok I guess if you plan to use the event emitter only once like is the case for the terminal now but I would like to use a more generic way. Or provide a `once` method in the emitter.

Reply:
Also I think like you said before it's more clean to call dispose and be certain the process is stopped.

Reply:
Ho and I really need that event handler for the internal kill() call too

Reply:
> Dispose would still be necessary to cleanup other external event emitters

They should be automatically disposed together with emitters if emitters are disposed in `handleOn*` methods.

It feels a bit weird that one should call `process.dispose` after `exit` event.

Reply:
>They should be automatically disposed together with emitters if emitters are disposed in handleOn* methods.

I don't get what you mean here handleOn methods can't dipose of the emitter if they did no other listeners would be called ?

And calling `process.dispose` after exit I think is fine since you may want to keep the Process object reference even after exit.  an exit event should not destroy the Process object.

Reply:
ok, I will say the PR looks then